clash-prelude/src/Clash/XException/MaybeX.hs

```diff
@@ -0,0 +1,149 @@
+Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+
+Helpers to make 'Clash.XException.XException' explicit in the type system.
+Using these helpers can help programmers account for 'Clash.XException.XException's
+properly in blackbox models or tests. Note that none of these operations can be
+translated to HDL.
+
+-}
+
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PatternSynonyms #-}
+
+module Clash.XException.MaybeX
+  ( MaybeX(IsX, IsDefined)
+
+    -- * Construction
+  , toMaybeX
+  , hasXToMaybeX
+
+    -- * Deconstruction
+  , unsafeToA
+
+    -- * Operations
+  , andX
+  , orX
+  , maybeX
+  ) where
+
+import Prelude
+
+import Control.Applicative
+import Control.DeepSeq (NFData)
+import Control.Exception (throw)
+
+import Clash.XException (XException(..), isX, hasX)
+
+-- | Structure helping programmers to deal with 'Clash.XException.XException'
+-- values. For safety reasons it can't be constructed directly, but should be
+-- constructed using either 'pure' or 'toMaybeX'. After construction, it can be
+-- deconstructed using either 'IsX' or 'IsDefined'.
+data MaybeX a
+  = IsX_ String
+  -- ^ Upon construction, @a@ evaluated to 'Clash.XException.XException'
+  | IsDefined_ !a
+  -- ^ Upon construction, @a@ evaluated to a non-bottom WHNF
+
+instance Show a => Show (MaybeX a) where
+  showsPrec d = \case
+    IsX_ msg     -> showParen (d > 10) $ showString "IsX "       . showsPrec 11 msg
+    IsDefined_ a -> showParen (d > 10) $ showString "IsDefined " . showsPrec 11 a
+
+-- | Upon construction, @a@ evaluated to 'Clash.XException.XException'
+pattern IsX :: forall a. String -> MaybeX a
+pattern IsX msg <- IsX_ msg
+
+-- | Upon construction, @a@ evaluated to a non-bottom WHNF
+pattern IsDefined :: forall a. a -> MaybeX a
+pattern IsDefined a <- IsDefined_ a
+{-# COMPLETE IsX, IsDefined #-}
+
+instance Functor MaybeX where
+  fmap _f (IsX_ msg) = IsX_ msg
+  fmap f  (IsDefined_ a) = either IsX_ IsDefined_ (isX (f a))
+
+instance Applicative MaybeX where
+  pure = either IsX_ IsDefined_ . isX
+
+  liftA2 f (IsDefined_ a) (IsDefined_ b) = IsDefined_ (f a b)
+  liftA2 _ (IsX_ msg)     _              = IsX_ msg
+  liftA2 _ _              (IsX_ msg)     = IsX_ msg
+
+-- | Construct a 'MaybeX' value. If @a@ evaluates to 'Clash.XException.XException',
+-- this function will return 'IsX'. Otherwise, it will return 'IsDefined'.
+toMaybeX :: a -> MaybeX a
+toMaybeX = pure
+
+-- | Construct a 'MaybeX' value. If 'hasX' evaluates to 'Left', this function
+-- will return 'IsX'. Otherwise, it will return 'IsDefined'.
+hasXToMaybeX :: NFData a => a -> MaybeX a
+hasXToMaybeX = either IsX_ IsDefined_ . hasX
+
+-- | Deconstruct 'MaybeX' into an @a@ - the opposite of 'toMaybeX'. Note that this
+-- is unsafe, because it might return an 'Clash.XException.XException' if the
+-- argument was 'IsX'. Use of this function should be accompanied with a comment
+-- explaining why it is fine to use.
+unsafeToA :: MaybeX a -> a
+unsafeToA = maybeX (throw . XException) id
```

Comment:
As discussed, don't forget to `fromMaybeX` and adjust the doc to be not so imposing ;-)